packages/lodestar/src/network/attestation-service/service.ts

```diff
@@ -0,0 +1,384 @@
+  }
+
+  /**
+   * Unsubscribe to a committee subnet from subscribedCommitteeSubnets.
+   * If a random subnet is present, we do not unsubscribe from it.
+   */
+  private handleUnsubscriptions(currentSlot: Slot): void {
+    const inactiveSubscribedCommitteeSubnets = this.subscribedCommitteeSubnets.getInactive(currentSlot);
+    const activeRandomSubnets = this.randomSubnets.getActive(currentSlot);
+    const fork = this.subscribedCommitteeSubnets.getForkName();
+    for (const subnet of inactiveSubscribedCommitteeSubnets) {
+      if (!activeRandomSubnets.includes(subnet)) {
+        this.gossip.unsubscribeTopic({type: GossipType.beacon_attestation, fork, subnet});
+      }
+    }
+    const inactiveCommitteeSubnets = this.committeeSubnets.getInactive(currentSlot);
+    // inactiveCommitteeSubnets is superset of inactiveSubscribedCommitteeSubnets
+    for (const subnet of inactiveCommitteeSubnets) {
+      this.committeeSubnets.delete(subnet);
+      this.subscribedCommitteeSubnets.delete(subnet);
+    }
+  }
+
+  /**
+   * A random subnet has expired.
+   * This function selects a new subnet to join, or extends the expiry if there are no more
+   * available subnets to choose from.
+   */
+  private handleRandomSubnetExpiry(slot: Slot): void {
+    const allRandomSubnets = this.randomSubnets.getAll();
+    const inactiveRandomSubnets = this.randomSubnets.getInactive(slot);
+    const currentFork = this.chain.getForkName();
+    const randomSubnetFork = this.randomSubnets.getForkName();
+    if (allRandomSubnets.length === ATTESTATION_SUBNET_COUNT) {
+      // We are at capacity, simply increase the timeout of the current subnet
+      for (const subnet of inactiveRandomSubnets) {
+        // After the fork occurs, let the subnets from the previous fork reach the end of life with no replacements
+        if (currentFork === randomSubnetFork) {
+          this.randomSubnets.request({
+            subnetId: subnet,
+            toSlot: getSubscriptionSlotForRandomSubnet(this.config, slot),
+          });
+        } else {
+          this.pruneRandomSubnet(subnet, slot);
+        }
+      }
+      return;
+    }
+    for (const subnet of inactiveRandomSubnets) {
+      this.pruneRandomSubnet(subnet, slot);
+      // After the fork occurs, let the subnets from the previous fork reach the end of life with no replacements
+      if (currentFork === randomSubnetFork) {
+        this.subscribeToRandomSubnet(currentFork);
+      }
+    }
+  }
+
+  /**
+   * A known validator has not sent a subscription in a while. They are considered offline and the
+   * beacon node no longer needs to be subscribed to the allocated random subnets.
+   *
+   * We don't keep track of a specific validator to random subnet, rather the ratio of active
+   * validators to random subnets. So when a validator goes offline, we can simply remove the
+   * allocated amount of random subnets.
+   */
+  private handleKnownValidatorExpiry(currentSlot: Slot): void {
```

Comment:
You could re-org this function by doing
1. Iterate `this.knownValidators` and delete expired directly without storing them on a set
2. Then compute `targetRandomSubnetCount = this.knownValidators.size * this.config.params.RANDOM_SUBNETS_PER_VALIDATOR`
3. Then if `activeRandomSubnets < targetRandomSubnetCount` add more, if `activeRandomSubnets > targetRandomSubnetCount` remove some

Reply:
Points 2 + 3 could be a new function `rebalanceRandomSubnets` and be called too after adding a new known validator, deduplicating logic